src/function/vector_hash_functions.cpp

```diff
@@ -10,199 +10,225 @@ namespace kuzu {
-                for (auto i = 0u; i < operandSelVector.getSelSize(); i++) {
-                    if (!operand.isNull(i)) {
-                        Hash::operation(operand.getValue<OPERAND_TYPE>(i), resultValues[i],
-                            &operand);
-                    } else {
-                        result.setValue(i, NULL_HASH);
-                    }
-                }
-            } else {
-                for (auto i = 0u; i < operandSelVector.getSelSize(); i++) {
-                    auto pos = operandSelVector[i];
-                    if (!operand.isNull(pos)) {
-                        Hash::operation(operand.getValue<OPERAND_TYPE>(pos), resultValues[pos],
-                            &operand);
-                    } else {
-                        resultValues[pos] = NULL_HASH;
-                    }
+            for (auto i = 0u; i < operandSelectVec.getSelSize(); i++) {
+                auto operandPos = operandSelectVec[i];
+                auto resultPos = resultSelectVec[i];
+                if (!operand.isNull(operandPos)) {
+                    Hash::operation(operand.getValue<OPERAND_TYPE>(operandPos),
+                        resultValues[resultPos]);
+                } else {
+                    result.setValue(resultPos, NULL_HASH);
                 }
             }
         }
     }
 }
 
-static std::unique_ptr<ValueVector> computeDataVecHash(ValueVector* operand) {
+template<typename LEFT_TYPE, typename RIGHT_TYPE, typename RESULT_TYPE, typename FUNC>
+static void executeOnValue(const common::ValueVector& left, common::sel_t leftPos,
+    const common::ValueVector& right, common::sel_t rightPos, common::ValueVector& result,
+    common::sel_t resultPos) {
+    FUNC::operation(left.getValue<LEFT_TYPE>(leftPos), right.getValue<RIGHT_TYPE>(rightPos),
+        result.getValue<RESULT_TYPE>(resultPos));
+}
+
+static void validateSelState(const common::SelectionVector& leftSelVec,
+    const common::SelectionVector& rightSelVec, const common::SelectionVector& resultSelVec) {
+    auto leftSelSize = leftSelVec.getSelSize();
+    auto rightSelSize = rightSelVec.getSelSize();
+    auto resultSelSize = resultSelVec.getSelSize();
+    if (leftSelSize > 1 && rightSelSize > 1) {
+        KU_ASSERT(leftSelSize == rightSelSize);
+        KU_ASSERT(leftSelSize == resultSelSize);
+    } else if (leftSelSize > 1) {
+        KU_ASSERT(leftSelSize == resultSelSize);
+    } else if (rightSelSize > 1) {
+        KU_ASSERT(rightSelSize == resultSelSize);
+    }
+}
+
+template<typename LEFT_TYPE, typename RIGHT_TYPE, typename RESULT_TYPE, typename FUNC>
+void BinaryHashFunctionExecutor::execute(const common::ValueVector& left,
+    const common::SelectionVector& leftSelVec, const common::ValueVector& right,
+    const common::SelectionVector& rightSelVec, common::ValueVector& result,
+    const common::SelectionVector& resultSelVec) {
+    validateSelState(leftSelVec, rightSelVec, resultSelVec);
```

Comment:
We should probably wrap this in KU_ASSERT so that it is not checked in release mode.